src/main/java/org/jenkinsci/remoting/nio/FifoBuffer.java

```diff
@@ -250,16 +251,19 @@ public int send(WritableByteChannel ch) throws IOException {
                     }
                     return 0;   // no data to read
                 }
+                try {
+                    int sent = r.send(ch, chunk);  // bytes actually written
 
-                int sent = r.send(ch, chunk);  // bytes actually written
-
-                read += sent;
-                sz -= sent;
+                    read += sent;
+                    sz -= sent;
 
-                lock.notifyAll();
+                    lock.notifyAll();
 
-                if (sent==0)    // channel filled up
-                    return read;
+                    if (sent == 0)    // channel filled up
+                        return read;
+                } catch (ClosedChannelException e) {
+                    return -1; // propagate EOF
```

Comment:
It looks like there is a potential for `read > 0` and so there actually was something sent in a previous pass of the loop, so returning `-1` would be a lie in that case?